File: NSTouchBarCatalogCreatingandCustomizingNSTouchBar/Objective-C/NSTouchBar Catalog/BackgroundWindow/BackgroundImagesViewController.m
/*
 Copyright (C) 2017 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 View controller used for the view-based table view of images, using a NSScrubberImageItemView without subclassing.
 */

#import "BackgroundImagesViewController.h"
#import "BackgroundViewController.h"
#import "TitleBarAccessoryViewController.h"
#import "ScrubberViewController.h"
#import "PhotoManager.h"

static NSTouchBarCustomizationIdentifier ScrubberCustomizationIdentifier =
    @"com.TouchBarCatalog.backgroudWindowScrubber";
static NSTouchBarItemIdentifier ScrubberItemIdentifier =
    @"com.TouchBarCatalog.backgroudWindowScrubberItem";

@interface BackgroundImagesViewController () <NSTouchBarDelegate, NSScrubberDelegate, NSScrubberDataSource, PhotoManagerDelegate>

@property (nonatomic, weak) IBOutlet NSScrollView *scrollView;
@property (nonatomic, weak) IBOutlet NSTableView *tableView;
@property (nonatomic, weak) IBOutlet NSProgressIndicator *progressIndicator;

@property (nonatomic, strong) IBOutlet NSArrayController *myContentArray;
@property (nonatomic, strong) NSMutableArray *tableContents;

@end


#pragma mark -

@implementation BackgroundImagesViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    // Listen for table view selection changes.
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(selectionDidChange:)
                                                 name:NSTableViewSelectionDidChangeNotification
                                               object:self.tableView];
    
    self.scrollView.wantsLayer = TRUE;
    self.scrollView.layer.cornerRadius = 6;
    
    // Load the pictures for our scrubber content.
    [self fetchPictureResources];
}

- (void)viewDidAppear
{
    [super viewDidAppear];
    
    [self.view.window makeFirstResponder:self.view];
}

- (void)displayPhotos {
    [self willChangeValueForKey:@"tableContents"];
    _tableContents = PhotoManager.shared.photos;
    [self didChangeValueForKey:@"tableContents"];
    
    self.touchBar = nil;   // Force update our NSTouchBar.
    
    [self.progressIndicator stopAnimation:self];
    self.progressIndicator.hidden = YES;
    self.scrollView.hidden = NO;
}

// Loads all the Desktop Pictures on this system, to be used for the image-based NSScrubber in the touch bar item, and for table view in the NSPopover.
- (void)fetchPictureResources
{
    if ([PhotoManager shared].loadComplete) {
        [self displayPhotos];
    } else {
        // The PhotoManager has not loaded all the photos. This could take a while, show the progress indicator.
        PhotoManager.shared.delegate = self;  // So we can be notified when the photos have been loaded.
        self.progressIndicator.hidden = false;
        self.scrollView.hidden = true;
        [self.progressIndicator startAnimation:self];
    }
}


#pragma mark NSTouchBarProvider

- (NSTouchBar *)makeTouchBar
{
    NSTouchBar *bar = [[NSTouchBar alloc] init];
    bar.delegate = self;
    
    bar.customizationIdentifier = ScrubberCustomizationIdentifier;
    
    // Set the default ordering of items.
    bar.defaultItemIdentifiers = @[ScrubberItemIdentifier, NSTouchBarItemIdentifierOtherItemsProxy];
    
    bar.customizationAllowedItemIdentifiers = @[ScrubberItemIdentifier];
    
    bar.principalItemIdentifier = ScrubberItemIdentifier;
    
    return bar;
}


#pragma mark - NSScrubberDataSource

static NSString *imageScrubberItemIdentifier = @"thumbnailItem";

- (NSInteger)numberOfItemsForScrubber:(NSScrubber *)scrubber
{
    return self.tableContents.count;
}

// Scrubber is asking for a custom view represention for a particuler item index.
- (NSScrubberItemView *)scrubber:(NSScrubber *)scrubber viewForItemAtIndex:(NSInteger)index
{
    NSScrubberImageItemView *itemView = [scrubber makeItemWithIdentifier:@"thumbnailItem" owner:nil];
    if (index < self.tableContents.count)
    {
        NSDictionary *itemDict = self.tableContents[index];
        itemView.image = [itemDict valueForKey:kImageKey];
    }
    
    itemView.imageView.imageScaling = NSImageScaleProportionallyDown;
    itemView.imageAlignment = NSImageAlignCenter;
    
    return itemView;
}


#pragma mark - NSScrubberFlowLayoutDelegate

// Scrubber is asking for the size for a particular item.
- (NSSize)scrubber:(NSScrubber *)scrubber layout:(NSScrubberFlowLayout *)layout sizeForItemAtIndex:(NSInteger)itemIndex
{
    return NSMakeSize(50, 30);
}


#pragma mark - NSScrubberDelegate

// User selected an image from the NSScrubber touch bar item.
- (void)scrubber:(NSScrubber *)scrubber didSelectItemAtIndex:(NSInteger)selectedIndex
{
    [self chooseImageWithIndex:selectedIndex];
}


#pragma mark NSTouchBarDelegate

// This gets called while the NSTouchBar is being constructed, for each NSTouchBarItem to be created.
- (nullable NSTouchBarItem *)touchBar:(NSTouchBar *)touchBar makeItemForIdentifier:(NSTouchBarItemIdentifier)identifier
{
    if ([identifier isEqualToString:ScrubberItemIdentifier])
    {
        // Create the scrubber touch bar item that uses the Desktop Pictures.
        NSCustomTouchBarItem *scrubberItem = [[NSCustomTouchBarItem alloc] initWithIdentifier:ScrubberItemIdentifier];
        
        NSScrubber *scrubber = [[NSScrubber alloc] initWithFrame:NSMakeRect(0, 0, 320, 30)];
        scrubber.delegate = self;   // This is so we can respond to selection.
        scrubber.dataSource = self; // This is so we can determine the content.
        
        [scrubber registerClass:[NSScrubberImageItemView class] forItemIdentifier:imageScrubberItemIdentifier];
            
        // For the image scrubber, we want the control to draw a fade effect to indicate that there is additional unscrolled content.
        scrubber.showsAdditionalContentIndicators = YES;

        scrubber.selectedIndex = 0; // Always select the first item in the scrubber.
        
        scrubberItem.customizationLabel = NSLocalizedString(@"Choose Photo", @"");
        
        NSScrubberLayout *scrubberLayout = [[NSScrubberFlowLayout alloc] init];
        scrubber.scrubberLayout = scrubberLayout;

        scrubber.mode = NSScrubberModeFree;
        scrubber.selectionBackgroundStyle = nil;
        scrubber.selectionOverlayStyle = nil;
        
        // Set the layout constraints on this scrubber so that it's 400 pixels wide.
        NSDictionary *items = NSDictionaryOfVariableBindings(scrubber);
        NSArray *theConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"H:[scrubber(400)]" options:0 metrics:nil views:items];
        [NSLayoutConstraint activateConstraints:theConstraints];
        
        scrubberItem.view = scrubber;
        
        return scrubberItem;
    }
    
    return nil;
}

- (void)chooseImageWithIndex:(NSInteger)imageIndex
{
    // Process the chosen image and dismiss us as the popover.
    TitleBarAccessoryViewController *presentingViewController = (TitleBarAccessoryViewController *)self.presentingViewController;
    BackgroundViewController *backgroundViewController = (BackgroundViewController *)presentingViewController.view.window.contentViewController;
        
    NSDictionary *itemDict = self.tableContents[imageIndex];
    
    NSURL *fileURL = [itemDict valueForKey:kImageURLKey];
    NSImage *fullImage = [[NSImage alloc] initWithContentsOfURL:fileURL];
    backgroundViewController.imageView.image = fullImage;
    
    [self dismissViewController:self];
    presentingViewController.openingViewController = nil;
}


#pragma mark - Notifications

// -------------------------------------------------------------------------------
//  Listens for changes table view row selection
// -------------------------------------------------------------------------------
- (void)selectionDidChange:(NSNotification *)notification
{
    NSInteger selectedRow = [self.tableView selectedRow];
    if (selectedRow != -1)
    {
        [self chooseImageWithIndex:selectedRow];
    }
}

#pragma mark - PhotoManagerDelegate

- (void)didLoadPhotos:(NSArray *)photos {
    [self displayPhotos];
}

@end

File: NSTouchBarCatalogCreatingandCustomizingNSTouchBar/Swift/NSTouchBar Catalog/BackgroundImagesViewController.swift
/*
 Copyright (C) 2017 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 View controller used for the view-based table view of images, using a NSScrubberImageItemView without subclassing.
 */

import Cocoa

@available(OSX 10.12.2, *)
class BackgroundImagesViewController: NSViewController {
    
    static let imageScrubber = NSTouchBarItem.Identifier("com.TouchBarCatalog.TouchBarItem.imageScrubber")

    var selectedItemIdentifier: NSTouchBarItem.Identifier = BackgroundImagesViewController.imageScrubber
    
    @IBOutlet weak var scrollView: NSScrollView!
    @IBOutlet weak var tableView: NSTableView!
    @IBOutlet weak var progressIndicator: NSProgressIndicator!
    
    @IBOutlet weak var myContentArray: NSArrayController!
    @objc var tableContents = [Any]()
    
    // MARK: View Controller Life Cycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Listen for table view selection changes.
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(BackgroundImagesViewController.selectionDidChange(_:)),
                                               name: NSTableView.selectionDidChangeNotification,
                                               object: tableView)
        
        scrollView.wantsLayer = true
        scrollView.layer?.cornerRadius = 6
        
        // Load the pictures for our scrubber and table content.
        fetchPictureResources()
    }
    
    override func viewDidAppear() {
        super.viewDidAppear()
        
        view.window?.makeFirstResponder(view)
    }
    
    fileprivate func displayPhotos() {
        willChangeValue(forKey: "tableContents")
        tableContents = PhotoManager.shared.photos
        self.didChangeValue(forKey: "tableContents")
        
        self.touchBar = nil    // Force update our NSTouchBar.
        
        self.progressIndicator.stopAnimation(self)
        self.progressIndicator.isHidden = true
        self.scrollView.isHidden = false
    }
    
    fileprivate func fetchPictureResources() {
        if PhotoManager.shared.loadComplete {
            displayPhotos()
        } else {
            // The PhotoManager has not loaded all the photos. This could take a while, show the progress indicator.
            PhotoManager.shared.delegate = self   // So we can be notified when the photos have been loaded.
            progressIndicator.isHidden = false
            scrollView.isHidden = true
            progressIndicator.startAnimation(self)
        }
    }
    
    // MARK: Notifications
    
    fileprivate func chooseImageWithIndex(index: Int) {
        guard let imageDict = tableContents[index] as? [String: Any] else { return }
        
        // Process the chosen image and dismiss us as the popover.
        if let presentingViewController = presenting as? TitleBarAccessoryViewController {
            if let backgroundViewController = presentingViewController.view.window?.contentViewController as? BackgroundViewController {
                if let url = imageDict[PhotoManager.ImageURLKey] as? NSURL {
                    if let fullImage = NSImage(contentsOf: url as URL) {
                        backgroundViewController.imageView.image = fullImage
                        presentingViewController.dismissViewController(self)
                        presentingViewController.openingViewController = nil
                    }
                }
            }
        }
    }
    
    @objc
    func selectionDidChange(_ notification: Notification) {
        // User selected a particular background photo.
        guard 0...tableView.numberOfRows ~= tableView.selectedRow else { return }
        
        chooseImageWithIndex(index: tableView.selectedRow)
    }
    
    static let itemViewIdentifier = "ImageItemViewIdentifier"
    
    // MARK: NSTouchBarProvider
    
    override func makeTouchBar() -> NSTouchBar? {
        let touchBar = NSTouchBar()
        touchBar.delegate = self
        touchBar.customizationIdentifier = NSTouchBar.CustomizationIdentifier("com.TouchBarCatalog.scrubberBar")
        touchBar.defaultItemIdentifiers = [BackgroundImagesViewController.imageScrubber]
        touchBar.customizationAllowedItemIdentifiers = [BackgroundImagesViewController.imageScrubber]
        touchBar.principalItemIdentifier = BackgroundImagesViewController.imageScrubber
        
        return touchBar
    }
}

// MARK: - NSTouchBarDelegate

@available(OSX 10.12.2, *)
extension BackgroundImagesViewController: NSTouchBarDelegate {
    
    // This gets called while the NSTouchBar is being constructed, for each NSTouchBarItem to be created.
    func touchBar(_ touchBar: NSTouchBar, makeItemForIdentifier identifier: NSTouchBarItem.Identifier) -> NSTouchBarItem? {
        let scrubberItem: NSCustomTouchBarItem
        
        scrubberItem = NSCustomTouchBarItem(identifier: identifier)
        scrubberItem.customizationLabel = NSLocalizedString("Choose Photo", comment:"")
        
        let scrubber = NSScrubber()
        scrubber.register(NSScrubberImageItemView.self,
                          forItemIdentifier: NSUserInterfaceItemIdentifier(BackgroundImagesViewController.itemViewIdentifier))
        scrubber.mode = .free
        scrubber.selectionBackgroundStyle = .roundedBackground
        scrubber.delegate = self
        scrubber.dataSource = self
        scrubber.showsAdditionalContentIndicators = true
        scrubber.scrubberLayout = NSScrubberFlowLayout()
        
        scrubberItem.view = scrubber
        
        // Set the scrubber's width to be 400.
        let viewBindings: [String: NSView] = ["scrubber": scrubber]
        let hconstraints = NSLayoutConstraint.constraints(withVisualFormat: "H:[scrubber(400)]",
                                                          options: [],
                                                          metrics: nil,
                                                          views: viewBindings)
        NSLayoutConstraint.activate(hconstraints)
        
        return scrubberItem
    }
}

// MARK: - PhotoManagerDelegate

@available(OSX 10.12.2, *)
extension BackgroundImagesViewController: PhotoManagerDelegate {

    // Used to refresh our UI, when all the photos have been loaded.
    func didLoadPhotos(photos: [Any]) {
        displayPhotos()
    }
}

// MARK: - NSScrubberDataSource

@available(OSX 10.12.2, *)
extension BackgroundImagesViewController: NSScrubberDataSource, NSScrubberDelegate {
    
    func numberOfItems(for scrubber: NSScrubber) -> Int {
        if tableContents.isEmpty {
            fetchPictureResources()
        }
        return tableContents.count
    }
    
    // Scrubber is asking for a custom view represention for a particuler item index.
    func scrubber(_ scrubber: NSScrubber, viewForItemAt index: Int) -> NSScrubberItemView {
        var returnItemView = NSScrubberItemView()
        if let itemView =
            scrubber.makeItem(withIdentifier: NSUserInterfaceItemIdentifier(rawValue: BackgroundImagesViewController.itemViewIdentifier),
                              owner: nil) as? NSScrubberImageItemView {
            if index < tableContents.count {
                if let imageDict = tableContents[index] as? [String: Any] {
                    if let image = imageDict[PhotoManager.ImageKey] as? NSImage {
                        itemView.image = image
                    }
                }
            }
            returnItemView = itemView
        }
        return returnItemView
    }
    
    // Scrubber is asking for the size for a particular item.
    func scrubber(_ scrubber: NSScrubber, layout: NSScrubberFlowLayout, sizeForItemAt itemIndex: Int) -> NSSize {
        return NSSize(width: 50, height: 30)
    }
    
    // User chose a particular image inside the scrubber.
    func scrubber(_ scrubber: NSScrubber, didSelectItemAt index: Int) {
        chooseImageWithIndex(index: index)
    }
}

